python/cudf/cudf/pylibcudf_tests/conftest.py

```diff
@@ -20,31 +23,124 @@
+
+            def _generate_list_data(typ):
+                child_colnames = []
+                if isinstance(typ, pa.ListType):
+                    # recurse to get vals
+                    rand_arrs, grandchild_colnames = _generate_list_data(
+                        typ.value_type
+                    )
+                    pa_array = pa.array(
+                        [list(row_vals) for row_vals in zip(rand_arrs)],
+                        type=typ,
+                    )
+                    child_colnames.append(("", grandchild_colnames))
+                else:
+                    # typ is scalar type
+                    pa_array = pa.array(rand_vals).cast(typ)
+                    child_colnames.append(("", []))
+                return pa_array, child_colnames
+
+            rand_arr, child_colnames = _generate_list_data(typ)
+        elif isinstance(typ, pa.StructType):
+
+            def _generate_struct_data(typ):
+                child_colnames = []
+                if isinstance(typ, pa.StructType):
+                    # recurse to get vals
+                    rand_arrs = []
+                    for i in range(typ.num_fields):
+                        rand_arr, grandchild_colnames = _generate_struct_data(
+                            typ.field(i).type
+                        )
+                        rand_arrs.append(rand_arr)
+                        child_colnames.append(
+                            (typ.field(i).name, grandchild_colnames)
+                        )
+
+                    pa_array = pa.StructArray.from_arrays(
+                        [rand_arr for rand_arr in rand_arrs],
+                        names=[
+                            typ.field(i).name for i in range(typ.num_fields)
+                        ],
+                    )
+                else:
+                    # typ is scalar type
+                    pa_array = pa.array(rand_vals).cast(typ)
+                return pa_array, child_colnames
+
+            rand_arr, child_colnames = _generate_struct_data(typ)
+        else:
+            rand_arr = pa.array(rand_vals).cast(typ)
+
+        table_dict[f"col_{typ}"] = rand_arr
+        colnames.append((f"col_{typ}", child_colnames))
+
+    pa_table = pa.Table.from_pydict(table_dict)
+
+    return plc.io.TableWithMetadata(
+        plc.interop.from_arrow(pa_table), column_names=colnames
+    ), pa_table
+
+
+@pytest.fixture(
+    params=["a.txt", pathlib.Path("a.txt"), io.BytesIO(), io.StringIO()],
+)
+def source_or_sink(request, tmp_path):
+    fp_or_buf = request.param
+    if isinstance(fp_or_buf, str):
+        fp_or_buf = f"{tmp_path}/{fp_or_buf}"
+    elif isinstance(fp_or_buf, os.PathLike):
+        fp_or_buf = tmp_path.joinpath(fp_or_buf)
+
+    yield fp_or_buf
+    # Cleanup after ourselves
+    # since the BytesIO and StringIO objects get cached by pytest
+    if isinstance(fp_or_buf, io.IOBase):
+        fp_or_buf.seek(0)
+        fp_or_buf.truncate(0)
```

Comment:
Wah?! This is "once-only" fixture, no, so it should be cleaned up automatically. What is this for?

Reply:
I couldn't figure out how to reset the StringIO buffers 😅 .
(otherwise we have the writes from the previous calls to write_json in it)

I tried setting the scope to "function", but that didn't seem to do anything.

Reply:
It should be the case that pytest conses a new object every time this fixture is used, so it seems something is wrong :(

Reply:
Can you distill this to a smaller test case, because this sounds absolutely like a pytest bug.

Reply:
Ah ok, yes, I was blind to this. This is just standard python, because pytest only sees the actual objects already.

The right way to handle this is something like:

```
@pytest.fixture(params=["string", "pathlib.Path", "BytesIO", "StringIO"])
def source_or_sink(request, tmp_path):
    if request.param == "string":
        return f"{tmp_path}/{request.param}"
    elif request.param == "pathlib.Path":
        return pathlib.Path(tmp_path) / request.param
    elif request.param == "BytesIO":
        return io.BytesIO()
    elif request.param == "StringIO":
        return io.StringIO()
    assert False
```